README.md
>ℹ️&nbsp;&nbsp;SignalFx was acquired by Splunk in October 2019. See [Splunk SignalFx](https://www.splunk.com/en_us/investor-relations/acquisitions/signalfx.html) for more information.

# Go client library for SignalFx

[![GoDoc](https://godoc.org/github.com/signalfx/signalfx-go?status.svg)](https://godoc.org/github.com/signalfx/signalfx-go)

This is a programmatic interface in Go for SignalFx's metadata and ingest APIs.

# SignalFlow

There is an **experimental** SignalFlow client in the `signalflow` directory.  An
example of its use is in [signalflow/example].  For full documentation see the
[godocs](https://godoc.org/github.com/signalfx/signalfx-go/signalflow).

# Example

```
import "github.com/signalfx/signalfx-go"

// The client can be customized by backing options onto the end. Check the
// godoc for more info!

// Instantiate your own client if you want to customize its options
// or test with a RoundTripper
httpClient := &http.Client{…}
client := signalfx.NewClient("your-token-here", HTTPClient(httpClient))

// Then do things!
chart, err := client.GetChart("abc123IdHere")
```

# Questions

## Why are there some things missing?

We're working on it, feel free to file an issue if an endpoint is missing!

## Why are the class names sometimes long and the source file names prefixed with `model_`?

The request and response bodies for this library are machine generated from our OpenAPI specs using [OpenAPI code generator](https://github.com/OpenAPITools/openapi-generator). This is a real boon for everyone, keeping the documentation as a source of truth and ensuring that this library has support for all the things!

This means that some of our type names are verbose. It's fine, you only type code once and the benefits are worth it.


CHANGELOG.md
# 1.26.0 13 Jan 2023

* Add `customMetricTypeDomains` field to the `GCPIntegration` struct [#180](https://github.com/signalfx/signalfx-go/pull/180)

# 1.25.0 22 Nov 2022

* Update list of Azure services [#175](https://github.com/signalfx/signalfx-go/pull/175), [#176](https://github.com/signalfx/signalfx-go/pull/176)

# 1.24.0 14 Oct 2022

* Add `importGCPMetriccs` flag to the `GCPIntegration` struct [#172](https://github.com/signalfx/signalfx-go/pull/172)
* Change type of `importAzureMonitor` flag in the `AzureIntegration` struct from bool to *bool [#172](https://github.com/signalfx/signalfx-go/pull/172)

# 1.23.0 27 Jul 2022

* Add logs list chart options to the model options [#168](https://github.com/signalfx/signalfx-go/pull/168)

# 1.22.0, 18 Jul 2022

* Add `useMetricSourceProjectForQuota` flag to the `GcpIntegration` struct [#167](https://github.com/signalfx/signalfx-go/pull/167)

# 1.21.0, 12 Jul 2022

* Add `importAzureMonitor` flag to the `AzureIntegration` struct [#165](https://github.com/signalfx/signalfx-go/pull/165)

# 1.19.0, 26 Apr 2022

* Remove `omitempty` from `ShowDataMarkers` field in the `Detector/Visualization` struct [#162](https://github.com/signalfx/signalfx-go/pull/162)

# 1.18.0, 12 Apr 2022

* Add `syncCustomNamespacesOnly` field to the `AwsCloudWatchIntegration` struct [#160](https://github.com/signalfx/signalfx-go/pull/160)

# 1.17.0, 1 Apr 2022

No functional changes, this is an equivalent of `1.12.0`.

We are releasing it to succeed mistakenly published `1.16.13` instead of `1.6.13`, which since Jan 2020 shows as the latest.

# 1.12.0, 17 Mar 2022

Add support for ServiceNow integration [#159]

# 1.11.0, 16 Mar 2022

* Add `IncludeList` field to `GCPIntegration` struct [#157](https://github.com/signalfx/signalfx-go/pull/157) and stop using deprecated `Whitelist` field

# 1.10.0, 14 Mar 2022

* Add `AdditionalServices` and `ResourceFilterRules` fields to the `AzureIntegration` struct [#156](https://github.com/signalfx/signalfx-go/pull/156)

# 1.9.0, 11 Mar 2022

* Add support for read permissions for dashboard and dashboard groups

* Add `AWS/S3/Storage-Lens` to AWS services list. Sorted AWS/Azure/GCP services by name. [#154](https://github.com/signalfx/signalfx-go/pull/154)

* Add `PollRateMs` (type `int64`, value in milliseconds) to define Azure and GCP integration poll rate [#153](https://github.com/signalfx/signalfx-go/pull/153) [#155](https://github.com/signalfx/signalfx-go/pull/155)

`PollRate` (type `PollRate`) is deprecated. Please use `PollRateMs` instead.<BR>
`PollRateMs` accepts any value between 1 and 10 minutes (`60000` - `600000`).

# 1.8.9, 3 Mar 2022

* signalflow: Fix race in computation buffering when closing (previously the messages were not getting fully flushed to the downstream channels) [#151](https://github.com/signalfx/signalfx-go/pull/151)

# 1.8.8, 28 Feb 2022

* Add MetricStreamsSyncState to AWS Cloudwatch integration [#147](https://github.com/signalfx/signalfx-go/pull/147)

# 1.8.6, 8 Feb 2022

* Adding AuthScopes attribute to the org token struct [#146](https://github.com/signalfx/signalfx-go/pull/146)

# 1.8.5, 27 Jan 2022

* signalflow fakebackend support for timestamp control and END_OF_CHANNEL [#145](https://github.com/signalfx/signalfx-go/pull/145)

# 1.8.4, 24 Jan 2022

* Adding ValidateDetector and its unit tests [#142](https://github.com/signalfx/signalfx-go/pull/142)

# 1.8.3, 17 Nov 2021

* Add newly supported AWS services to the list. [#140](https://github.com/signalfx/signalfx-go/pull/140)

# 1.8.2, 30 Sep 2021

*  Update dependencies [#135](https://github.com/signalfx/signalfx-go/pull/135)

# 1.8.1, 24 Aug 2021

* Context cancellation handling [#131](https://github.com/signalfx/signalfx-go/pull/131)
* Allow to run PostDisconnectCallback

# 1.8.0, 2021-05-11

* Don't send empty orderBy query parameter for metric metadata calls [#127](https://github.com/signalfx/signalfx-go/pull/127).
* Replace nonfunctioning `RetrieveMetricMetadataResponseModel.Result` field with `Results` [#127](https://github.com/signalfx/signalfx-go/pull/127).

# 1.7.18, 2021-03-29

* Fix handling of client base URLs that have paths in them

# 1.7.17, 2021-03-23

* Update list of Azure services [#123](https://github.com/signalfx/signalfx-go/pull/123)

# 1.7.15, 2021-02-10

* Added `TimeZone` property to Detector. [#119](https://github.com/signalfx/signalfx-go/pull/119)

# 1.7.14, 2021-01-28

* Return events/alerts from signalflow events/alerts queries

# 1.7.13, 2020-12-07

*  signalflow: Fix channel end/abort control message handling

# 1.7.12, 2021-02-10

* Added `hideMissingValues` to chart.Options to show or hide missing values in the chart. [#111](https://github.com/signalfx/signalfx-go/pull/111)
* Added `minDelay` field to detector.
* Added `PagerDutyIntegrationGetByName` method.

# 1.7.11, 2020-10-19

* Fix processing ControlMessage in signalflow.Computation

# 1.7.10, 2020-08-06
* Don't omit chart.description from JSON, as API doesn't clear it if absent. [#107](https://github.com/signalfx/signalfx-go/pull/107)

# 1.7.9, 2020-08-06
* Allow AWS poll rate to be an int64. [#106](https://github.com/signalfx/signalfx-go/pull/106)

# 1.7.8, 2020-08-06
* Added `detectorName` to incident model. Thanks [choo-stripe](https://github.com/choo-stripe)! [#103](https://github.com/signalfx/signalfx-go/pull/103)
* Clean up some resources when a context is canceled during Signalflow computations. Thanks [kerbyhughes](https://github.com/kerbyhughes)! [#104](https://github.com/signalfx/signalfx-go/pull/104)
* Add new team linking methods for detectors and dashboard groups. [#102](https://github.com/signalfx/signalfx-go/pull/102)
* Add new team fields. [#105](https://github.com/signalfx/signalfx-go/pull/105)

# 1.7.7, 2020-07-22
* Added new fields to Azure integration.
* Fix race condition in SignalFlow client Close operation that could cause
  panics.

# 1.7.5, 2020-06-11
* Fix busted accessor.

# 1.7.4, 2020-06-11

client.go
package signalfx

import (
	"context"
	"io"
	"net/http"
	"net/url"
	stdpath "path"
	"time"

	"github.com/signalfx/signalfx-go/signalflow"
)

// DefaultAPIURL is the default URL for making API requests
const DefaultAPIURL = "https://api.signalfx.com"

// AuthHeaderKey is the HTTP header used to pass along the auth token
// Note that while HTTP headers are case insensitive this header is case
// sensitive on the tests for convenience.
const AuthHeaderKey = "X-Sf-Token"

// Client is a SignalFx API client.
type Client struct {
	baseURL    string
	httpClient *http.Client
	authToken  string
	userAgent  string
}

// ClientParam is an option for NewClient. Its implementation borrows
// from Dave Cheney's functional options API
// (https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis).
type ClientParam func(*Client) error

// NewClient creates a new SignalFx client using the specified token.
func NewClient(token string, options ...ClientParam) (*Client, error) {
	client := &Client{
		baseURL: DefaultAPIURL,
		httpClient: &http.Client{
			Timeout: time.Second * 30,
		},
		authToken: token,
		userAgent: "signalfx-go",
	}

	for _, option := range options {
		option(client)
	}

	return client, nil
}

// APIUrl sets the URL that our client will communicate with, allowing
// it to be adjusted to another URL for testing or communication with other
// SignalFx clusters. Example `"https://api.signalfx.com"`.
func APIUrl(apiURL string) ClientParam {
	return func(client *Client) error {
		client.baseURL = apiURL
		return nil
	}
}

// UserAgent sets the UserAgent string to include with the request.
func UserAgent(userAgent string) ClientParam {
	return func(client *Client) error {
		client.userAgent = userAgent
		return nil
	}
}

// HTTPClient sets the `http.Client` that this API client will use to
// to communicate. This allows you to replace the client or tune it to your
// needs.
func HTTPClient(httpClient *http.Client) ClientParam {
	return func(client *Client) error {
		client.httpClient = httpClient
		return nil
	}
}

func (c *Client) doRequest(ctx context.Context, method string, path string, params url.Values, body io.Reader) (*http.Response, error) {
	return c.doRequestWithToken(ctx, method, path, params, body, c.authToken)
}

func (c *Client) doRequestWithToken(ctx context.Context, method string, path string, params url.Values, body io.Reader, token string) (*http.Response, error) {
	destURL, err := url.Parse(c.baseURL)
	if err != nil {
		return nil, err
	}
	destURL.Path = stdpath.Join(destURL.Path, path)

	if params != nil {
		destURL.RawQuery = params.Encode()
	}
	req, err := http.NewRequestWithContext(ctx, method, destURL.String(), body)
	if token != "" {
		req.Header.Set(AuthHeaderKey, token)
	}
	req.Header.Set("User-Agent", c.userAgent)
	req.Header.Set("Content-Type", "application/json")
	if err != nil {
		return nil, err
	}

	return c.httpClient.Do(req)
}

// Added temporarily to support navigator API. Must be worked on before the change gets made.
func (c *Client) doRequestWithNavHeader(ctx context.Context, method string, path string, params url.Values, body io.Reader) (*http.Response, error) {
	return c.doRequestWithTokenWithNavHeader(ctx, method, path, params, body, c.authToken)
}

// Added temporarily to support navigator API. Must be worked on before the change gets made.
func (c *Client) doRequestWithTokenWithNavHeader(ctx context.Context, method string, path string, params url.Values, body io.Reader, token string) (*http.Response, error) {
	destURL, err := url.Parse(c.baseURL)
	if err != nil {
		return nil, err
	}
	destURL.Path = stdpath.Join(destURL.Path, path)

	if params != nil {
		destURL.RawQuery = params.Encode()
	}
	req, err := http.NewRequestWithContext(ctx, method, destURL.String(), body)
	if token != "" {
		req.Header.Set(AuthHeaderKey, token)
	}
	req.Header.Set("User-Agent", c.userAgent)
	req.Header.Set("Content-Type", "application/vnd.splunk.observability.navigator+json")
	if err != nil {
		return nil, err
	}

	return c.httpClient.Do(req)
}

// SignalFlow creates and returns a SignalFlow client that can be used to
// execute streaming jobs.
func (c *Client) SignalFlow(options ...signalflow.ClientParam) (*signalflow.Client, error) {
	options = append(options, signalflow.AccessToken(c.authToken))
	return signalflow.NewClient(options...)
}
